enhancements/insights/pulling-pull-secret-from-ocm.md

```diff
@@ -0,0 +1,168 @@
+5. If there is a mismatch, the Insights operator applies the new `pull-secret`.
+   * Note: MCO updating the `pull-secret` on the nodes will not result in node reboots.
+6. The Insights operator itself starts using the new `pull-secret`
+7. Cluster comes up with the new `pull-secret` and starts reporting back to OCM into the new account.
+
+Note: This process will work only on the clusters where the feature was released as part of the Insights operator and will not be backported to older versions.
+
+### User Stories
+
+As an OCM user, I would like to transfer cluster ownership from my account to another user's account in one step.
+
+### API Extensions
+
+No API extensions.
+
+### Risks and Mitigations
+
+Risk: The OCM API is down or returning HTTP response code >= 500.
+
+Mitigation: The Insights Operator is marked as degraded in such case.
+
+Risk: The `pull-secret` does not work or is otherwise corrupted.
+
+Mitigation: We must refer to the OCP documentation, which describes the manual fix procedure. 
+
+
+## Design Details
+
+### Authorization
+
+The Insights Operator is able to pull the updated `pull-secret` content from the OCM API using the existing `cloud.openshift.com` token
+available in the `pull-secret` (in the `openshift-config` namespace).
+
+The Insights Operator must provide a cluster ID as an identifier of the cluster.
+
+### OCM API
+
+The OCM API provides an endpoint that will serve the complete `pull-secret` definition when the migration window is opened. 
+The migration window will be opened for certain period of time (e.g 5 days) so tha the Insights Operator can safely update the secret.
+
+Open question: What exactly will the API return outside of the given migration window?  
+
+### Monitoring
+
+There will be a new metric registered by the Insights Operator, which will count the number of responses (according the HTTP response status code) from the OCM API.
+
+
+### Test Plan
+
+- `insights-operator-e2e-tests` suite can verify that the Insights Operator is able to download and update the `pull-secret` content
+
+### Graduation Criteria
+
+Thera are no graduation criteria. This feature is planned as a fast track feature.
+
+#### Dev Preview -> Tech Preview
+
+Empty
+
+#### Tech Preview -> GA
+
+Empty
+
+#### Removing a deprecated feature
+
+The periodical data pulling can be easily disabled in the cluster configuration. Removing this feature will require updating the Insights operator code base.
+
+### Upgrade / Downgrade Strategy
+
+There is no upgrade/downgrade strategy needed.
+
+### Version Skew Strategy
+
+There is no Skew strategy needed. This work should have no impact on the upgrade. It doesn't require any coordinated behavior in the control plane. No other components will change.
+
+
+### Operational Aspects of API Extensions
+
+No API extension.
+
+#### Failure Modes
+
+TBD
+
+#### Support Procedures
+
+TBD
```

Comment:
@mrunalp, @aravindhp What about these two chapters? Do we need to mention anything? WDYT pls?

Reply:
I think we need to call out how we are going to handle scenarios where the cluster is not able to update the secret in the cluster during Window for whatever reason. Maybe send an event so that the user is aware?

Reply:
Alerting, in both a public Insights rule and an on-cluster warning alert escalating to critical in the last day or something would be the loudest shouting for "the window is closing, didn't you want this handoff?!".  Might be worth giving folks a n OCM-side way to abort the handoff to silence the alarms if they'd changed their minds, although I guess OCM-UX is out-of-scope here.

Reply:
I think we won't be able to count the days within the window, but I think we can count the number of unsuccessful responses within the window and then fire an alert when a threshold is reached. WDYT?

Reply:
OK so the API will provide time information so we should be able to inform user about the last day or something...

Reply:
I would resolve this conversation thread, but what about the `Failure Modes` chapter?

Reply:
for Failure Modes we just dream up and discuss as many potential issues as we can think of:

* Some `paused` compute pool fails to pick up the new secret and continues attempting to use the old secret.  No problems if the old secret keeps working, but if it is revoked... FIXME: MCO/workloads folks might be able to clarify what will happen.
* Admin never confirms the handoff.  In this case, the window expires; no issues in-cluster.  If they open a support case "I thought I did transfer that over", support will `${STEPS}` to get logs showing that the transfer was initiated, alerted upon, but never confirmed in cluster. (possibly by visiting Telemetry, or maybe this would be summarized in an Insights rule?)
* Attacker who can forge an X.509 cert and impersonate OCM (or who gains server-side access to OCM) recommends an invalid pull secret.  Do we attempt to guard against this with some sort of pre-apply validation checks?  Do we mark it down as unlikely and have a generic "recovering your cluster from a broken pull secret" section coupled with our "remove `${KEY}` from your pull secret to tell Insights to stop polling upstream" section?
* More?

Reply:
Yep. Good points. Thanks.

- ad `#1` - We must mention the fact that the old secret will work for some time. Who can help us with the clarification (on the MCO side) please?
- ad `#2` - There's no such confirmation planned. We've discussed that the Insights operator is unable to provide such confirmation, because the "real" update is then up to the MCO. So we rely on Telemetry here (+ optionally an Insights rule)
- ad `#3` - Can we do some pre-apply validation @aravindhp? I don't think so, but it's definitely worth mentioning.

Reply:
I don't think you need to worry about MCO application for handoff confirmation.  The Insights operator knows and will alert on "pending transition waiting for local admin approval" per your existing proposal, and the Insights operator also knows if it accepts the approval and updates the local pull secret (regardless of whether the MCO then rolls that secret out).  So it should be possible to create an Insights rule that points out the most recent handoff attempt and then whether that handoff went through or not.  But I'm fine if the position is that, while a rule is possible, the risk of confusion is small enough that it's not worth writing the rule up, and we can always re-evaluate later if support bumps into this often enough that we want to revisit our guess at the risk size.

Reply:
Yes. Thanks. I hopefully mentioned all the points here. Resolving the conversation.